cli/shared/src/main/scala/com/lightbend/rp/reactivecli/docker/DockerRegistry.scala

```diff
@@ -52,84 +52,95 @@ object DockerRegistry extends LazyLogging {
       Failure(new IllegalArgumentException(s"""Cannot parse uri "$uri"""))
     else
       Success(
         Image(
           url = providedUrl.getOrElse(DockerDefaultRegistry),
           namespace = providedNs.getOrElse(DockerDefaultLibrary),
           image = image,
-          tag = providedTag.getOrElse(DockerDefaultTag),
+          ref = providedRef.getOrElse(ImageTag(DockerDefaultTag)),
           providedUrl = providedUrl,
           providedNamespace = providedNs,
           providedImage = image,
-          providedTag = providedTag))
+          providedRef = providedRef))
   }
 
   private def getBlob(http: HttpExchange, credentials: Option[HttpRequest.Auth], useHttps: Boolean, validateTls: Boolean)(img: Image, digest: String, token: Option[HttpRequest.BearerToken]): Future[(HttpResponse, Option[HttpRequest.BearerToken])] =
     for {
       r <- getWithAuth(http, credentials, validateTls, blobUrl(img, digest, useHttps), HttpHeaders(Map.empty), token, Some(img.pullScope))
     } yield r
 
-  private def getTags(http: HttpExchange, credentials: Option[HttpRequest.Auth], useHttps: Boolean, validateTls: Boolean)(img: Image, token: Option[HttpRequest.BearerToken])(implicit settings: HttpSettings): Future[(Option[String], Option[HttpRequest.BearerToken])] =
+  private def checkRepositoryValid(http: HttpExchange, credentials: Option[HttpRequest.Auth], useHttps: Boolean, validateTls: Boolean)(img: Image, token: Option[HttpRequest.BearerToken])(implicit settings: HttpSettings): Future[(Either[String, Unit], Option[HttpRequest.BearerToken])] =
     for {
-      r <- getWithAuth(http, credentials, validateTls, tagsUrl(img, useHttps), HttpHeaders(Map()), token, Some(img.pullScope))
-    } yield r._1.body -> r._2
-
-  private def imgValid(tags: Option[String], img: Image): Try[Boolean] = {
-    tags match {
-      case None =>
-        Failure(new IllegalArgumentException(s"got unexpected docker registry response"))
-      case Some(str) =>
-        val tags = Parse.parseOption(str).flatMap(_.hcursor.downField("tags").focus)
-        val foundTag = tags.flatMap(j =>
-          j.hcursor.downArray.find(tag => tag.isString && tag.string == Some(img.tag)).focus)
-        (tags, foundTag) match {
-          case (None, None) => Failure(new IllegalArgumentException(s"image doesn't seem to exist in docker registry"))
-          case (Some(_), None) => Failure(new IllegalArgumentException(s"image ${img.image} doesn't have tag named ${img.tag}"))
-          case _ => Success(true)
-        }
-    }
-  }
+      // We only fetch a single tag (?n=1) because we only care about status codes, and this saves data transfer for
+      // large repositories.
+
+      r <- getWithAuth(http, credentials, validateTls, s"${tagsUrl(img, useHttps)}?n=1", HttpHeaders(Map()), token, Some(img.pullScope))
+    } yield (
+      r._1.statusCode match {
+        case x if x >= 200 && x <= 299 => Right(())
+        case 404 => Left("Unable to find repository or registry")
```

Comment:
Nitpick: error message should start with lowercase letter, since it is concatenated into sentence like "Failed to obtain Docker config for img, Unable to find repository or registry"

Reply:
Good catch.